File: C++/042. Trapping Rain Water.cpp
class Solution {
public:
    int trap(vector<int>& height) 
    {
        int n = height.size();
        if (n <= 2) return 0;
        vector<int> left = vector<int>(n, 0);
        vector<int> right = vector<int>(n, 0);
        left[0] = height[0];
        right[n-1] = height[n-1];
        for (int i = 1; i < n; i++)
        {
            left[i] = max(left[i-1], height[i]);
        }
        for (int j = n - 2; j >= 0; j--)
        {
            right[j] = max(right[j+1], height[j]);
        }
        int res = 0;
        for (int k = 1; k < n - 1; k++)
        {
            res += max(min(left[k-1], right[k+1]) - height[k], 0);
        }
        return res;
    }
};

File: python/042. Trapping Rain Water.py
class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        the collected rain at position x: max(min(leftBars[x-1],right[x+1]) - height[x], 0)
        leftBars[x]: the highest bar between 0 and x
        rightBars[x]: the highest bar between x and n-1
        """
        n = len(height)
        if n < 3: return 0
        leftBars = [0] * n
        rightBars = [0] * n
        leftBars[0] = height[0]
        rightBars[n-1] = height[n-1]
        for i in range(1, n):
            leftBars[i] = max(leftBars[i-1], height[i])
        for i in range(n-2, -1, -1):
            rightBars[i] = max(rightBars[i+1], height[i])
        res = 0
        for k in range(1, n-1):
            res += max(0, min(leftBars[k-1], rightBars[k+1]) - height[k])
        return res